lib/triton_kernel_call.cc

```diff
@@ -0,0 +1,129 @@
+/* Copyright 2022 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "triton_kernel_call.h"
+
+#include <iostream>
+#include <string>
+
+#include <pybind11/pybind11.h>
+#include "cuda.h"
+
+namespace py = pybind11;
+
+namespace jax_triton {
+
+void TritonExecutable::launch(CUstream stream, void** buffers) {
+  CUdevice dev;
+  CUcontext ctx;
+  // Set the current context to the stream context so we can query the stream
+  // device
+  cuStreamGetCtx(stream, &ctx);
+  cuCtxSetCurrent(ctx);
+  /// Only load the kernel if it hasn't already been loaded for this device
+  cuCtxGetDevice(&dev);
+  if (!is_loaded(dev)) {
+    load(dev);
+  } 
+  CUfunction kernel = kernels[dev];
+  std::string params;
+  params.resize(8 * arity);
+  char* params_ptr = &params[0];
+  for (uint32_t i = 0; i < arity; i++) {
+    params_ptr = (char*)(((uintptr_t)params_ptr + 7) & (-8));
+    std::memcpy(params_ptr, &buffers[i], 8);
+    params_ptr += 8;
+  }
+  size_t params_size = (std::ptrdiff_t)(params_ptr - &params[0]);
+  void* config[] = {
+    CU_LAUNCH_PARAM_BUFFER_POINTER,
+    static_cast<void*>(const_cast<char*>(params.data())),
+    CU_LAUNCH_PARAM_BUFFER_SIZE, &params_size,
+    CU_LAUNCH_PARAM_END
+  };
+  CUresult result = cuLaunchKernel(kernel, grid_0, grid_1, grid_2, num_warps * 32, 1, 1, shared_mem, stream, nullptr, config);
+  if (result != 0) {
+    std::cout << "Failed launch: " << result << std::endl;
+  }
+};
+
+void TritonExecutable::load(CUdevice device) {
+  const std::lock_guard<std::mutex> lock(mut);
+  // Mimics Triton kernel loading
+  std::string assembly;
+  if (asm_map.find("cubin") != asm_map.end())
+    assembly = py::cast<std::string>(asm_map["cubin"]);
+  else
```

Comment:
Add an assertion that asm_map must have "ptx" in the else branch?

Reply:
Done.